File: src/channels/webchat.ts
import { BaseChannel, NormalisedMessage } from "./base";

export class WebChatChannel extends BaseChannel {
  readonly channelName = "webchat";

  async sendMessage(_recipientId: string, _text: string): Promise<void> {
    throw new Error("WebChatChannel.sendMessage requires a live WebSocket/SSE connection.");
  }

  receiveMessage(payload: unknown): NormalisedMessage {
    const p = payload as Record<string, unknown>;
    return {
      senderId: String(p["user_id"] ?? ""),
      text: String(p["message"] ?? ""),
      channel: this.channelName,
      raw: payload,
    };
  }
}

File: openclaw_safe/channels/webchat.py
"""WebChat channel adapter – generic HTTP/WebSocket-based chat interface."""

from __future__ import annotations

from typing import Any

from .base import BaseChannel


class WebChatChannel(BaseChannel):
    """Adapter for a browser-based WebChat widget."""

    channel_name = "webchat"

    def send_message(self, recipient_id: str, text: str, **kwargs: Any) -> None:
        # In production: push message to client via WebSocket or SSE
        raise NotImplementedError(
            "WebChatChannel.send_message requires a live WebSocket/SSE connection."
        )

    def receive_message(self, payload: dict[str, Any]) -> dict[str, Any]:
        return {
            "sender_id": payload.get("user_id", ""),
            "text": payload.get("message", ""),
            "channel": self.channel_name,
            "raw": payload,
        }
